controlplane/kubeadm/internal/etcd/etcd.go

```diff
@@ -0,0 +1,233 @@
+	ClientURLs []string
+
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+
+	// Alarms is the list of alarms for a member.
+	Alarms []AlarmType
+}
+
+// pbMemberToMember converts the protobuf representation of a cluster member to a Member struct.
+func pbMemberToMember(m *etcdserverpb.Member) *Member {
+	return &Member{
+		ID:         m.GetID(),
+		Name:       m.GetName(),
+		PeerURLs:   m.GetPeerURLs(),
+		ClientURLs: m.GetClientURLs(),
+		IsLearner:  m.GetIsLearner(),
+		Alarms:     []AlarmType{},
+	}
+}
+
+// NewEtcdClient creates a new etcd client with a custom dialer and is configuration with optional functions.
+func NewEtcdClient(endpoint string, dialer GRPCDial, tlsConfig *tls.Config) (*clientv3.Client, error) {
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{endpoint},
+		DialTimeout: etcdTimeout,
+		DialOptions: []grpc.DialOption{
+			grpc.WithBlock(), // block until the underlying connection is up
+			grpc.WithContextDialer(dialer),
+		},
+		TLS: tlsConfig,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create etcd client")
+	}
+	etcdClient.Endpoints()
+	return etcdClient, nil
+}
+
+// NewClientWithEtcd configures our response formatter (Client) with an etcd client and endpoint.
+func NewClientWithEtcd(etcdClient etcd) (*Client, error) {
+	if len(etcdClient.Endpoints()) == 0 {
+		return nil, errors.New("etcd client was not configured with any endpoints")
+	}
+	return &Client{
+		Endpoint:   etcdClient.Endpoints()[0],
+		EtcdClient: etcdClient,
+	}, nil
+}
+
+// Close closes the etcd client.
+func (c *Client) Close() error {
+	return c.EtcdClient.Close()
+}
+
+// Member retrieves the Member information for a given peer.
+func (c *Client) Member(ctx context.Context, peerURL string) (*Member, error) {
+	members, err := c.Members(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, m := range members {
+		if m.PeerURLs[0] == peerURL {
+			return m, nil
+		}
+	}
+	return nil, errors.Errorf("etcd member %v not found", peerURL)
+}
+
+// Members retrieves a list of etcd members.
+func (c *Client) Members(ctx context.Context) ([]*Member, error) {
+	response, err := c.EtcdClient.MemberList(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get list of members for etcd cluster")
+	}
+
+	alarms, err := c.Alarms(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	members := make([]*Member, len(response.Members))
+	for i, m := range response.Members {
+		newMember := pbMemberToMember(m)
+		for _, c := range alarms {
+			if c.MemberID == newMember.ID {
+				newMember.Alarms = append(newMember.Alarms, c.Type)
+			}
+			members[i] = newMember
+		}
+	}
+
+	return members, nil
+}
+
+// Status returns the cluster status as known by the connected etcd server.
+func (c *Client) Status(ctx context.Context) (*clientv3.StatusResponse, error) {
```

Comment:
Did you mean to return the raw `clientv3.StatusResponse`? I think we would want a wrapper struct like `Member`.

Reply:
Let's remove it if it's unneeded. We can always add it back later if we need it.